remote/client.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path"
+
+	//log "github.com/coreos/flannel/Godeps/_workspace/src/github.com/golang/glog"
+	"github.com/coreos/flannel/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/flannel/subnet"
+)
+
+type httpRespErr struct {
+	resp *http.Response
+	err  error
+}
+
+func httpDo(ctx context.Context, req *http.Request) (*http.Response, error) {
+	// Run the HTTP request in a goroutine and pass the response to f.
+	tr := &http.Transport{}
+	client := &http.Client{Transport: tr}
+	c := make(chan httpRespErr, 1)
+	go func() {
+		resp, err := client.Do(req)
+		c <- httpRespErr{resp, err}
+	}()
+
+	select {
+	case <-ctx.Done():
+		tr.CancelRequest(req)
+		<-c // Wait for f to return.
+		return nil, ctx.Err()
+	case r := <-c:
+		return r.resp, r.err
+	}
+}
+
+// implements subnet.Manager by sending requests to the server
+type RemoteManager struct {
+	client http.Client
+	host   string // includes scheme, host, and port
+}
+
+func NewRemoteManager(listenAddr string) subnet.Manager {
+	return &RemoteManager{host: "http://" + listenAddr}
+}
+
+func (m *RemoteManager) mkurl(network string, parts ...string) string {
+	if network == "" {
+		network = "/_"
+	}
+	if network[0] != '/' {
+		network = "/" + network
+	}
+	return m.host + path.Join(append([]string{network}, parts...)...)
+}
+
+func (m *RemoteManager) GetNetworkConfig(ctx context.Context, network string) (*subnet.Config, error) {
```

Comment:
RemoteManager implements subnet.Manager interface which has `ctx` as the first arg for consistency. Not all methods/implementations (there're two now) do anything with it yet. New etcd client is ctx based so once we move over to it, it'll be used more.